### ISSUE TITLE
PoC WordPress-GitHub integration using Documents From Git

### ISSUE THREAD

**Issue by ty2k** [BCDevOps/developer-experience#2189]
**Describe the issue**
We would like to explore using WordPress as a solution for hosting our technical documentation. Using the GDX WordPress team's instance of WordPress means we wouldn't have to handle the complexity of hosting and managing the service, and could focus on the content.

[Documents from Git](https://wordpress.org/plugins/documents-from-git/) is a WordPress plugin for WordPress v5+ which is designed to generate WordPress pages from Markdown files hosted in git repositories.

Explore the feasibility of this plugin as a solution for our technical documentation.

**Additional context**
- Check for ability to integrate with the WordPress search engine. Ex: Users can run a search across WordPress content AND GitHub imported content at the same time.

**Definition of done**
- [ ] Install the Documents from Git plugin in <s>our [Private Cloud PaaS WordPress instance](https://platform-services-dev.apps.silver.devops.gov.bc.ca/) which is hosted by the GDX WordPress team</s> a local WordPress development instance.
- [ ] Configure the plugin to pull documentation from multiple GitHub documentation repositories, public and private.
- [ ] Document the feasibility of this plugin as a solution for pulling Markdown documentation from our GitHub repositories.

**Comment by ty2k** [comment]
# Documents from Git research

**TL;DR**: While Documents from Git is useful for adding pages to WordPress using Markdown documentation from git repositories, it would require a lot of WordPress administration activity and additional work to make it behave nicely for us. Documents must be added and managed one at a time through shortcodes. Images must use an absolute link or they won't work. The document cache must be manually busted by changing a time-to-live setting, or by setting a short interval for refreshes. :x:  The WordPress built-in search doesn't appear to index documentation contents.

This testing was performed locally on a [WordPress development instance](https://github.com/WordPress/wordpress-develop).

## Configuration

[Documents from Git](https://wordpress.org/plugins/documents-from-git/) gets installed in a WordPress instance through the WP Plugins page. The plugin uses GitHub's [Markdown API](https://docs.github.com/en/rest/reference/markdown) to do the work of turning Markdown files into HTML markup.

The plugin can be used without authenticating to GitHub, but when doing so, your requests are rate-limited (60 requests per minute).

To authenticate, from the Settings -> Documents from Git page, you provide the plugin with a GitHub username and [access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) that it will use to make requests. The plugin will now be using the individual user's access to both access GitHub repositories and parse Markdown documents. If a user has been granted access to a private repository, this repository can be accessed by Documents from Git. Per Olena, we would use this with a team robot account.

## Shortcodes

Individual Markdown documents get added by using [WordPress shortcodes](https://wordpress.com/support/shortcodes/) that reference a git hosting platform (GitHub, GitLab, etc) and an action (`markdown` to parse Markdown into HTML) along with a URL for the specific document you want to render. Example:

```
[git-github-markdown url="https://github.com/ty2k/bc-cms-prototype/blob/main/README.md"]
```

![WordPress Gutenberg editor with a Document from Git shortcode block](https://images.zenhubusercontent.com/5ec709daacce4878764898c4/d4a4288e-9669-4d06-b53e-7171bc3bf0a6)

![WordPress preview page showing a rendered Markdown block](https://images.zenhubusercontent.com/5ec709daacce4878764898c4/4e933f9f-7c8a-4ee2-a85a-de47736ad42f)

This method of referencing pages using shortcodes has pros and cons for us:

**Pros**

- A WordPress page can be composed of multiple Markdown file by referencing multiple shortcodes if desired.
- If a Markdown file is unavailable (gets renamed, moved, deleted, etc) when the plugin is trying to pull its source, the WordPress page that references it stays up, but the block referencing the Markdown file will display a 404 error. With the page staying up on the WordPress side, that gives us the chance to update the reference to the Markdown file, proactively re-direct the WordPress page, or something else.
- Using a shortcode with the `checkout` action, a label can be added to the WordPress page that links to the document on GitHub and shows the date it was last updated. In the existing DevHub model, we have to manually add a "Last updated" field in the Markdown file itself and then remember to update it to indicate when a document was last touched.
- Using a shortcode with the `history` action, a list of the latest commits to the document can be added (with the limit set by us). This could be useful for showing how the document has changed over time, in the case of rapidly changing (beta) documents.
- CSS customizations are made easy with surrounding wrapper classes that we can specify.

**Cons**

- WordPress shortcodes can be easily broken with typos, or by inexperienced users trying to add new codes. Using shortcodes means you now have code living in WordPress as a version control system, which is not ideal. The more shortcodes that get added over time, the more active management becomes necessary. 
- No special handling is included for frontmatter. When viewing Markdown files on GitHub, frontmatter is handled by displaying it in a `<table>` at the top of the document. In Documents from Git, it will be rendered as if it were pure Markdown (ex: a series of paragraphs or `<ul>`s between two `<hr>` dividers).
- Level 1 heading tags in the source Markdown files (lines preceded with a single `#`, conventionally at the top of a Markdown document or just after any frontmatter) will correctly get rendered as `<h1>` tags for display. But in WordPress, the page title is the `<h1>` heading. So a WordPress page with a page title that is importing a Markdown file with an `#` heading will have more than one `<h1>` tag. Having multiple `<h1>` tags on a page is allowed by the HTML specification, but it is not good for screenreader users who navigate using these tags (see MDN's [Multiple `<h1>` elements on one page](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Heading_Elements#multiple_h1_elements_on_one_page)).

## Caching

The caching strategy in Documents from Git will cache pages in the WordPress database, without a mechanism for auto-updating when a repository is updated. The cache allows you to specify a global time-to-live, so all documents can be set to update daily, weekly, etc.

The entire cache must be flushed if you desire a fresh copy of any single document using its latest state. There is currently no button to do this. At this time, you must go into the settings and change the time-to-live value in order to generate a new cache of all the documents managed by the plugin. Updating a single document could be awkward for us since it involves manual intervention. Any WordPress administrator can perform this manual cache-busting action.

Compare this process to current DevHub where a build of the app must be kicked off to do a fresh pull of documents from GitHub (in DevHub, documents are refreshed at build time, instead of at run time in WordPress).

## Images

Images that are referenced in Markdown files will only work using an absolute link. Relative image links work in GitHub, and this is a typical way of referencing images (ie. images get checked into the repository alongside documentation, and then a relative link is used so the images can be parsed as expected in any environment, like on GitHub itself or on a developer's machine). Existing documentation using a relative linking strategy will have broken images. Documentation that happens to use absolute links, like a hotlink to another repository or an image hosting service, should work as expected.

## Search :x:

WordPress's built-in search feature appears to only index the text of the shortcode itself, as opposed to the rendered contents of the documentation being referenced. We would need to build this functionality ourselves or use a different search service that does this.

## Plugin Metrics

Currently, Documents from Git is listed as having 300+ active installations in the WordPress plugin library. Its [GitHub repo](https://github.com/gis-ops/wordpress-markdown-git) has 9 watchers, 10 forks, and 32 stars. In my searching, I have not found a more widely used plugin that does something similar.
